swift3/controllers/cors.py

```diff
@@ -0,0 +1,180 @@
+
+    return resp
+
+
+def check_cors_rule(data):
+    '''Check at minima CORS rules'''
+    rules = data.findall('CORSRule')
+    for rule in rules:
+        origin = rule.find('AllowedOrigin')
+        if origin.text.count('*') > 1:
+            raise CORSInvalidRequest(
+                'AllowedOrigin "%s" can not have more than one wildcard'
+                % origin.text)
+
+        for method in rule.findall('AllowedMethod'):
+            if method.text not in CORS_ALLOWED_HTTP_METHOD:
+                raise CORSInvalidRequest(
+                    "Found unsupported HTTP method in CORS config. "
+                    "Unsupported method is %s" % method.text)
+        for exposed in rule.findall('ExposeHeader'):
+            if '*' in exposed.text:
+                raise CORSInvalidRequest(
+                    'ExposeHeader "%s" contains wildcard. We currently do '
+                    'not support wildcard for ExposeHeader.' % exposed.text)
+        for allowed in rule.findall('AllowedHeader'):
+            if allowed.text.count('*') > 1:
+                raise CORSInvalidRequest(
+                    'AllowedHeader "%s" can not have more than one wildcard.'
+                    % allowed.text)
+
+
+class CorsController(Controller):
+    """
+    Handles the following APIs:
+
+     - GET Bucket CORS
+     - PUT Bucket CORS
+     - DELETE Bucket CORS
+
+    """
+    @public
+    @bucket_operation
+    def GET(self, req):  # pylint: disable=invalid-name
+        """
+        Handles GET Bucket CORS.
+        """
+        resp = req._get_response(self.app, 'HEAD',
+                                 req.container_name, None)
+        body = resp.sysmeta_headers.get(BUCKET_CORS_HEADER)
+        if not body:
+            raise NoSuchCORSConfiguration
+        return HTTPOk(body=body, content_type='application/xml')
+
+    @public
+    @bucket_operation
+    def PUT(self, req):  # pylint: disable=invalid-name
+        """
+        Handles PUT Bucket CORs.
+        """
+        xml = req.xml(MAX_CORS_BODY_SIZE)
+        try:
+            data = fromstring(xml, "CorsConfiguration")
+        except (XMLSyntaxError, DocumentInvalid):
+            raise MalformedXML()
+        except Exception as e:
+            exc_type, exc_value, exc_traceback = sys.exc_info()
+            LOGGER.error(e)
+            raise exc_type, exc_value, exc_traceback
+
+        # forbid wildcard for ExposeHeader
+        check_cors_rule(data)
+
+        req.headers[BUCKET_CORS_HEADER] = xml
+        resp = req._get_response(self.app, 'POST',
+                                 req.container_name, None)
+        return resp
+
+    @public
+    @bucket_operation
+    def DELETE(self, req):  # pylint: disable=invalid-name
+        """
+        Handles PUT Bucket CORs.
+        """
+        req.headers[BUCKET_CORS_HEADER] = ''
+        return req._get_response(self.app, 'POST',
+                                 req.container_name, None)
```

Comment:
We may have to hack the response status: S3 expects 204, but swift returns 202 for POST requests. Same issue for PUT. See the TaggingController https://github.com/open-io/swift3/pull/19